### PR TITLE
PCHR-2535: Disallow transitions to non-cancelling statuses with a negative closing balance

### PR DESCRIPTION
**Pull request by igorpavlov**
## Overview

At the moment is it possible to:

1) Set Leave Request status to More information required with a negative closing balance by Manager
2) Set Leave Request status to Awaiting for approval with a negative closing balance by Staff
3) Approve a Leave Request with a negative closing balance by manager

This PR fixes this issue.

## Technical Details

The problem was in the function that calculates the balance change for **existing** leave requests:

```php
public static function getTotalBalanceChangeForLeaveRequest(LeaveRequest $leaveRequest, $expiredOnly = false) {
  ...
  return $balance; // for Leave Requests using current TOIL balance this value was negative, though the actual calculation was correct
}
```
This lead to the following condition to pass while it should have been failing:

```php
//                                        v That expression will always be false
if(!$absenceType->allow_overuse && $leaveRequestBalance > $currentBalance) {
  throw new InvalidLeaveRequestException(
    'There are only X days leave available...'
  );
}
```
This issue was fixed by simply returning the original (not absolute) value in the balance calculation function for **new** requests:

```php
private static function calculateBalanceChangeFromCreateParams($params) {
  // ...
  return $leaveRequestBalance['amount']; // <-- remove abs()
}
```

## Comments

This PR also fixes trailing spaces.

This PR also fixes the existing test, which ignored the calculation of balance change. The test was still passing, however, if we remove the implementation this test actually checks, it will throw another error (no working days specified), not the error we expect to be thrown (negative closing balance).

---

- [x] Tests Pass

### CODE REVIEW

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/LeaveBalanceChange.php [hunk 1]
```diff
@@ -405,7 +405,7 @@ public static function getTotalBalanceChangeForLeaveRequest(LeaveRequest $leaveR
       }
     }
 
-    return $balance;
+    return abs($balance);
```
Comment (tunbola) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/LeaveBalanceChange.php]: I think some tests usually require this function to return a negative balance and not positive.
And really the result should be negative because the leave requests cause a negative balance, if you need it in positive it should be in the function where this is returned.

Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/LeaveBalanceChange.php]: `getLeaveRequestBalance` returns `calculateBalanceChangeFromCreateParams()` or `getTotalBalanceChangeForLeaveRequest()`.

`calculateBalanceChangeFromCreateParams` returns **abs()**
but 
`getTotalBalanceChangeForLeaveRequest` doesn't.

Is that expected?

Comment (tunbola) [reply uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/LeaveBalanceChange.php]: Yeah its expected. If you check the code `calculateBalanceChangeFromCreateParams` returns an absolute value for the results returned by the `LeaveRequest::calculateBalanceChange` function, You need to wrap the value returned from  `getTotalBalanceChangeForLeaveRequest` in `abs` also as we need the leave balance in positive to compare it against the current entitlement balance.

Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/LeaveBalanceChange.php]: Yep, did exactly what you said ^^. Was not sure that's the best solution, but since you suggest the same, I am not much more confident. Thank you.

**Comment by davialexandre** [comment]
@igorpavlov can you try changing the condition [here](https://github.com/civicrm/civihr/blob/la-milestone-3/uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/BAO/LeaveRequest.php#L489)

to:

```php
!$absenceType->allow_overuse && ($currentBalance + $leaveRequestBalance) < 0
```

And see it that will fix the problem?

**Comment by tunbola** [comment]
In that case, @davi, IF we change to what you suggested, we need to change `calculateBalanceChangeFromCreateParams` to not return an absolute value anymore.
I actually like this better.

**Comment by igorpavlov** [comment]
@Davi thank you, I will certainly try. However, I have a slight concern, since `$leaveRequestBalance =self::getLeaveRequestBalance($params)` and `getLeaveRequestBalance()` can return negative values as well as positive depending on what is inside params.

**Comment by davialexandre** [comment]
@tunbola good point! And now I understand why we had this `$leaveRequestBalance > $currentBalance in the first place`

So here is an explanation of what happened and probably why we never spotted this problem before:
- On #2149 we changed that validation a little bit. Before that, we used to get the Leave Request balance with the `calculateBalanceChangeFromCreateParams` method, which always return a positive value.
- After #2149 we now use the `getLeaveRequestBalance()` method, but it is a bit inconsistent. For existing Leave Requests it will return a negative value (and the validation will fail) and for new Leave Request it will return a positive value (making the validation pass).

So what @igorpavlov solution was trying to do was to somehow reintroduce the behavior we had before #2149. Unfortunately, it caused some tests to fail.

What this validation is trying to do anyway is to check if the current balance + the leave request balance (that can be either positive or negative) will result in a negative balance. So that's I suggested the condition change. That plus what @tunbola suggested should fix the issue @igorpavlov

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php [hunk 2]
```diff
@@ -910,6 +910,219 @@ public function testManagerCanCancelOrRejectLeaveRequestEvenIfBalanceIsGreaterTh
...
+      'request_type' => LeaveRequest::REQUEST_TYPE_LEAVE
+    ];
+
+    $this->setExpectedException(
+      'CRM_HRLeaveAndAbsences_Exception_InvalidLeaveRequestException',
+      'There are only '. $entitlementBalance .' days leave available. This request cannot be made or approved'
+    );
+
+    // Create new request by staff with Awaiting Approval status
+    $leaveRequest = LeaveRequestFabricator::fabricateWithoutValidation($leaveRequestData, true);
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php]: It would be better to move this statement before the `setExpectedException()` one, just to make sure that the exception will be thrown by LeaveRequest::create().

This applies to the other tests

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php [hunk 3]
```diff
@@ -910,6 +910,219 @@ public function testManagerCanCancelOrRejectLeaveRequestEvenIfBalanceIsGreaterTh
...
+    // Approve leave request by manager
+    $updatedLeaveRequestData = $leaveRequestData;
+    $updatedLeaveRequestData['id'] = $leaveRequest->id;
+    $updatedLeaveRequestData['status_id'] = $leaveRequestStatuses['approved'];
+    $updatedLeaveRequest = LeaveRequest::create($updatedLeaveRequestData);
+
+    // Test that the status has not been changed
+    $this->assertEquals($updatedLeaveRequest->status_id, $leaveRequestStatuses['awaiting_approval']);
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php]: The test stops when it gets the exception it is expecting. This assertion will never be reached.

This applies to the other tests

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php [hunk 4]
```diff
@@ -910,6 +910,219 @@ public function testManagerCanCancelOrRejectLeaveRequestEvenIfBalanceIsGreaterTh
     }
   }
 
+  public function testManagerCanNotApproveLeaveRequestIfBalanceChangeIsMoreThanEntitlementBalanceWhenAllowOveruseFalse() {
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php]: The tests you added all mention a manager or a staff, but I don't think this is really necessary. The validation you're trying to test here doesn't care about the user role. At this point, we should only worry about the balances. 

What we need to test here is that an open leave request cannot be updated if its balance change will make the balance negative). In this case an open leave request can be either:
- An Approved Leave Request that was updated to either Awaiting approval or More Information Required
- An More Information Required that was updated to Awaiting Approval or vice versa

Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php]: @davialexandre so you mean it should be possible (for example) transit between More Information Required and Awaiting Approval even if the closing balance is negative?

**Comment by igorpavlov** [comment]
@Davi, final check:

@Tunbola wrote:

>In that case, @DAVI, IF we change to what you suggested, we need to change calculateBalanceChangeFromCreateParams to not return an absolute value anymore.
I actually like this better.

Wouldn't that affect the front-end?

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php (hunk 4 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php]: Sure. Just wondering, is such an approach allowed? It allows to test multiple transitions in one test:
```js
$statusTransitions = [
  ['approved', 'more_information_required'],
  ['awaiting_approval', 'approved']
];

foreach ($statusTransitions as $statusTransition) {
  $statusFrom = $statusTransition[0];
  $statusTo = $statusTransition[1];
  $exceptionMessage = '';

  // Balance change is 1 since this is a single day leave request
  $leaveRequestData = [
    'type_id' => $absenceType->id,
    'contact_id' => $periodEntitlement->contact_id,
    'status_id' => $statusFrom,
    'from_date' => $requestDate,
    'from_date_type' => $requestDateType,
    'to_date' => $requestDate,
    'to_date_type' => $requestDateType,
    'request_type' => LeaveRequest::REQUEST_TYPE_LEAVE
  ];

  // Create a new leave request
  $leaveRequest = LeaveRequestFabricator::fabricateWithoutValidation($leaveRequestData, true);

  // Attempt to change leave request status
  $updatedLeaveRequestData = $leaveRequestData;
  $updatedLeaveRequestData['id'] = $leaveRequest->id;
  $updatedLeaveRequestData['status_id'] = $statusTo;

  // Attempt to transit the leave request status
  try {
    $updatedLeaveRequest = LeaveRequest::create($updatedLeaveRequestData);
  } catch (Exception $e) {
    $exceptionMessage = $e->getMessage();
  }

  $this->assertEquals($exceptionMessage, 'There 2 are only '. $entitlementBalance .' days leave available. This request cannot be made or approved');
}
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php [hunk 5]
```diff
@@ -910,6 +910,71 @@ public function testManagerCanCancelOrRejectLeaveRequestEvenIfBalanceIsGreaterTh
...
+    HRJobContractFabricator::fabricate(
+      ['contact_id' => $periodEntitlement->contact_id],
+      ['period_start_date' => $periodStartDate]
+    );
+
+    $workPattern = WorkPatternFabricator::fabricateWithA40HourWorkWeek(['is_default' => 1]);
+    ContactWorkPatternFabricator::fabricate([
+      'contact_id' => $contactId,
+      'pattern_id' => $workPattern->id
+    ]);
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php]: Since you've fabricated the Work Pattern as the default one, it's not necessary to assign it to the contact using a `ContactWorkPattern`. Default Work Patterns are used whenever a contact doesn't have any patterns.

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php [hunk 6]
```diff
@@ -910,6 +910,71 @@ public function testManagerCanCancelOrRejectLeaveRequestEvenIfBalanceIsGreaterTh
...
+
+    $leaveRequestStatuses = array_flip(LeaveRequest::buildOptions('status_id', 'validate'));
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php]: This:
```php
array_flip(LeaveRequest::buildOptions('status_id', 'validate'));
```

Can be replaced with:

```php
LeaveRequest::getStatuses()
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php [hunk 7]
```diff
@@ -910,6 +910,71 @@ public function testManagerCanCancelOrRejectLeaveRequestEvenIfBalanceIsGreaterTh
...
+
+    // Balance change is 1 since this is a single day leave request
+    $leaveRequestData = [
+      'type_id' => $absenceType->id,
+      'contact_id' => $periodEntitlement->contact_id,
+      'status_id' => $leaveRequestStatuses['approved'],
+      'from_date' => $requestDate,
+      'from_date_type' => $requestDateType,
+      'to_date' => $requestDate,
+      'to_date_type' => $requestDateType,
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php]: The Leave Request fabricator has some default values for the `from_date_type` and `to_date_type` so you don't need to specify them here.

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php [hunk 8]
```diff
@@ -910,6 +910,71 @@ public function testManagerCanCancelOrRejectLeaveRequestEvenIfBalanceIsGreaterTh
...
+    ]);
+
+    $leaveRequestStatuses = array_flip(LeaveRequest::buildOptions('status_id', 'validate'));
+
+    // Balance change is 1 since this is a single day leave request
+    $leaveRequestData = [
+      'type_id' => $absenceType->id,
+      'contact_id' => $periodEntitlement->contact_id,
+      'status_id' => $leaveRequestStatuses['approved'],
+      'from_date' => $requestDate,
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php]: You're not using the `$requestDate` variable anywhere else in this code, so there's no need to keep this in a separate variable. Please change this to:

```php
'from_date' => CRM_Utils_Date::processDate('2016-06-14'),
```

Please also do the same to `to_date`

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php [hunk 9]
```diff
@@ -910,6 +910,71 @@ public function testManagerCanCancelOrRejectLeaveRequestEvenIfBalanceIsGreaterTh
...
+  public function testLeaveRequestCannotBeUpdatedIfBalanceChangeIsMoreThanEntitlementBalanceWhenAllowOveruseFalse() {
+    $contactId = 1;
+    $periodStartDate = CRM_Utils_Date::processDate('2016-01-01');
+    $periodEndDate = CRM_Utils_Date::processDate('2016-12-31');
+    $requestDate = CRM_Utils_Date::processDate('2016-06-14');
+    $requestDateType = $this->leaveRequestDayTypes['all_day']['value'];
+
+    $absencePeriod = AbsencePeriodFabricator::fabricate([
+      'start_date' => $periodStartDate,
+      'end_date'   => $periodEndDate,
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php]: You're not using these variables anywhere else. You can just pass in the dates here instead.

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php [hunk 10]
```diff
@@ -910,6 +910,71 @@ public function testManagerCanCancelOrRejectLeaveRequestEvenIfBalanceIsGreaterTh
...
+    $periodEntitlement = LeavePeriodEntitlementFabricator::fabricate([
+      'type_id' => $absenceType->id,
+      'contact_id' => $contactId,
+      'period_id' => $absencePeriod->id
+    ]);
+
+    // Current balance must be less than the balance change
+    $entitlementBalance = 0;
+    $this->createLeaveBalanceChange($periodEntitlement->id, $entitlementBalance);
+    $periodStartDate = $absencePeriod->start_date;
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php]: There's no need for this assignment here. You can just pass the `$absencePeriod->start_date` value to the Job Contract fabricator

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php (hunk 5 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php]: I don't even seem to need to fabricate a work pattern in this case.

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php (hunk 8 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php]: I use this var twice and I think it provides an understanding that this is a single day request:

``php
...
'from_date' => $requestDate,
'from_date_type' => $requestDateType,
'to_date' => $requestDate,
'to_date_type' => $requestDateType,
...
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php (hunk 7 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php]: I get an exception: 'The from_date_type field should not be empty'

Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/BAO/LeaveRequestTest.php]: Investigating...

**Comment by igorpavlov** [comment]
Just squashed my PR Feedbacks into one.
